### PR TITLE
Donate and reset password buttons

### LINKED ISSUE #3141: Donation button inside the game [Feature Request]
Put a link to https://supertuxkart.net/Donate somewhere inside the game, ideas:

-  Create a button named _Support_ or _Donate_ with an icon like the heart of the website(or any other icon), write a short description and put the donate link
- Put the link inside the _( i ) About_ screen

### PR DESCRIPTION
**Pull request by qwertychouskie**
I tested it on a few *buntu VMs and on Wine, and it worked, though it should be tested on macOS and real Windows before merging.  Closes #3141, #3160, #1946.

### CODE REVIEW

**Comment by deveee** [comment]
I'm not convinced to this feature on linux, because there is a lot window managers and browsers that may behave differently. For example if you have a window manager that doesn't support netwm, then still you will have displayed STK over your opened web browser. Or some broken window managers may change your resolution when you leave STK and not change it back after closing a browser etc...

I also found such topic:
https://www.reddit.com/r/androiddev/comments/2vsyq4/beware_google_bans_app_from_playstore_because_of/
so I'm not sure if adding donation link is good idea.

**Comment by Alayan-stk-2** [comment]
It would need to be disabled for Android then, but Google's selfish rules ("we want our share of your donations !") should certainly not decide what go into the main release.

**Comment by deveee** [comment]
Google is rather strict for applications on the play store :P But indeed ```#ifndef ANDROID``` is ok.

But anyway I think that leave fullscreen on linux is unsafe. I had a lot of issues in the past (for example game window was black when I returned to the game after alt+tab, on some older gnome versions it wasn't even possible to use alt+tab or move to secod desktop, because fullscreen window was displayed on top of everything etc.). Even now on ubuntu 18.04 nvidia blob enables panning for some reason when I set lower resolution in the game and it causes a mess. Plus the problem with netwm (some exotic WMs don't support it).

Anyway it's not that big problem, because nobody is forced to click on this link. I just can't guarantee that it will work for everyone.

File: data/gui/online/register.stkgui [hunk 1]
```diff
@@ -34,6 +34,8 @@
                     <textbox id="password" proportion="2" height="fit" I18N="In the registration dialog"/>
                 </div>
 
+                <!--<button id="password_reset" width="fit" text="Reset password"/>-->
```
Comment (auriamg) [inline data/gui/online/register.stkgui]: What is this commented line doing there?

File: src/online/link_helper.cpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+#ifdef _WIN32
+#include <windows.h>
+#include <shellapi.h>
+#endif
+
+using namespace Online;
+
+namespace Online
+{
+    void LinkHelper::OpenURL (std::string url) {
```
Comment (auriamg) [inline src/online/link_helper.cpp]: Braces go on a newline

File: src/online/link_helper.cpp [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+#ifdef _WIN32
+        ShellExecuteA(NULL, "open", url.c_str(), NULL, NULL, SW_SHOWNORMAL);
+#else 
+#ifdef APPLE
+        std::string op = std::string("open ").append(url);
+#else
+        std::string op = std::string("xdg-open ").append(url);
+#endif
+        system(op.c_str());
```
Comment (auriamg) [inline src/online/link_helper.cpp]: You should handle the return code and display an error message if an error code is returned. This is shaky as it is, error handling is not optional

File: src/states_screens/credits.cpp [hunk 4]
```diff
@@ -393,6 +394,10 @@ void CreditsScreen::eventCallback(GUIEngine::Widget* widget,
     {
         StateManager::get()->escapePressed();
     }
+    if (name == "donate")
+    {
+        Online::LinkHelper::OpenURL("https://SuperTuxKart.net/Donate"); // Open donation page
```
Comment (auriamg) [inline src/states_screens/credits.cpp]: This should preferably come from the stk_config.xml config file and not be hardcoded, I think? @hiker would you agree?

File: src/states_screens/register_screen.cpp [hunk 5]
```diff
@@ -428,6 +432,10 @@ void RegisterScreen::eventCallback(Widget* widget, const std::string& name,
             onEscapePressed();
         }
     }
+    else if (name == "password_reset")
+    {
+        Online::LinkHelper::OpenURL("http://addons.SuperTuxKart.net/password-reset.php"); // Open password reset page
```
Comment (auriamg) [inline src/states_screens/register_screen.cpp]: Same thing, should ideally come from a config file

File: src/states_screens/user_screen.cpp [hunk 6]
```diff
@@ -342,6 +346,10 @@ void BaseUserScreen::eventCallback(Widget* widget,
         }
         makeEntryFieldsVisible();
     }
+    else if (name == "password_reset")
+    {
+        Online::LinkHelper::OpenURL("http://addons.SuperTuxKart.net/password-reset.php"); // Open password reset page
```
Comment (auriamg) [inline src/states_screens/user_screen.cpp]: config file

**Comment by auriamg** [comment]
Help would be welcome to test this on various linux distributions. It's true that we probably can't get it to work everywhere, though if it works for most that would be acceptable

**Comment by nado** [comment]
Tested with xmonad, it opens correctly in the browser, the text on the buttons disappears when resized under a certain size.
It works the same in fullscreen or tiling : it is the size that matters.

Built with `cmake .. -DUSE_FRIBIDI=0 -DUSE_WIIUSE=0 -DCMAKE_BUILD_TYPE=Release`

**Comment by auriamg** [comment]
On hold until the requested changes have been applied

**Comment by deveee** [comment]
I can confirm that it works for me under Gnome. But please add the ```#if !defined(ANDROID)``` to avoid problems on google play. The ```xdg-open``` won't work on android anyway.

File: src/states_screens/register_screen.cpp (hunk 5 above)
Comment (auriamg) [reply src/states_screens/register_screen.cpp]: This still applies

File: src/states_screens/credits.cpp (hunk 4 above)
Comment (auriamg) [reply src/states_screens/credits.cpp]: This still applies

File: src/online/link_helper.cpp [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+            Log::error("OpenURL", "Command returned non-zero exit status");
+        }
+#elif defined(__linux__) && !defined(__ANDROID__)
+        std::string command = std::string("xdg-open ").append(url);
+        if (system(command.c_str()))
+        {
+            Log::error("OpenURL", "Command returned non-zero exit status");
+        }
+#else
+        Log::error("OpenURL", "Not implemented for this platform!");
```
Comment (auriamg) [inline src/online/link_helper.cpp]: In cases where it's not supported, the donate button should be hidden.No use having a button that does nothing when you click on it

File: src/states_screens/user_screen.cpp (hunk 6 above)
Comment (auriamg) [reply src/states_screens/user_screen.cpp]: Still applies

File: src/online/link_helper.cpp [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+using namespace Online;
+
+namespace Online
+{
+    void LinkHelper::OpenURL (std::string url)
+    {
+#if defined(_WIN32)
+        ShellExecuteA(NULL, "open", url.c_str(), NULL, NULL, SW_SHOWNORMAL);
+#elif defined(APPLE)
```
Comment (auriamg) [inline src/online/link_helper.cpp]: should be __APPLE__

**Comment by jpenguin** [comment]
Apple's Clang distribution can be identified by the presence of the __apple_build_version__ macro.
GCC & Intel C++ on OSX can be identified by the presence of the __APPLE__ && __MACH__ macro.

https://sourceforge.net/p/predef/wiki/browse_pages/

File: src/states_screens/credits.cpp [hunk 9]
```diff
@@ -205,7 +206,11 @@ void CreditsScreen::loadedFromFile()
         // translations should be just before the last screen
         m_sections.swap( m_sections.size() - 1, m_sections.size() - 2 );
     }
-
+#ifdef __ANDROID__
```
Comment (auriamg) [inline src/states_screens/credits.cpp]: The special case for Android is ok, but there should also be a similar check to disable the button for unsupported platforms. LinkHelper::isSupported() or somethinG like that

This is the donate button, but same thing for the reset password button

**Comment by qwertychouskie** [comment]
Thanks!

**Pull request merged**